File: backend/internal/domain/theme.go
package domain

import (
	"context"
	"time"

	"github.com/Voltage11/tplatform/internal/types/filterbool"
	"github.com/google/uuid"
)

type Theme struct {
	ID          uuid.UUID
	Name        string
	Description string
	IsActive    bool
	CreatedByID uuid.UUID
	CreatedAt   time.Time
	DateBegin   *time.Time
	DateEnd     *time.Time
	MaxPoint    int
	CheckPoint  int
	ImgPath     string
	DeletedAt   *time.Time
}

type ThemeWithDetail struct {
	Theme
	CreatedBy UserShort
}

type ThemeFilter struct {
	Name          string
	IsActive      filterbool.FilterBool
	CreatedByID   *uuid.UUID
	CreatedAtFrom *time.Time
	CreatedAtTo   *time.Time
	Pagination    PaginationRequest
}

// Repository Репозиторий для работы с темами
type ThemeRepository interface {
	GetByID(ctx context.Context, id uuid.UUID) (*Theme, error)
	GetByIDWithDetail(ctx context.Context, id uuid.UUID) (*ThemeWithDetail, error)
	Create(ctx context.Context, theme *Theme) error
	Update(ctx context.Context, theme *Theme) error
	Delete(ctx context.Context, id uuid.UUID) error
	SetActive(ctx context.Context, id uuid.UUID, isActive bool) error
	GetList(ctx context.Context, filter ThemeFilter) ([]*ThemeWithDetail, int64, error)
}

// Service Сервис для работы с темами
type ThemeService interface {
	GetByID(ctx context.Context, id uuid.UUID) (*Theme, error)
	GetByIDWithDetail(ctx context.Context, id uuid.UUID) (*ThemeWithDetail, error)
	Create(ctx context.Context, theme *Theme) error
	Update(ctx context.Context, theme *Theme) error
	Delete(ctx context.Context, id uuid.UUID) error
	SetActive(ctx context.Context, id uuid.UUID, isActive bool) error
	GetList(ctx context.Context, filter ThemeFilter) (*PagedResult[*ThemeWithDetail], error)
}

File: frontend/src/types/theme.ts
export interface Theme {
  id: string;
  name: string;
  description?: string | null;
  is_active: boolean;
  created_by_id: string;
  created_at: string;
  date_begin?: string | null;
  date_end?: string | null;
  max_point: number;
  check_point: number;
  img_path?: string | null;
  created_by: { id: string; name: string };
}

export interface ThemeCreate {
  name: string;
  description?: string | null;
  is_active: boolean;
  date_begin?: string | null;
  date_end?: string | null;
  check_point: number;
  img_path?: string | null;
}

export interface ThemeUpdate {
  name: string;
  description?: string | null;
  is_active: boolean;
  date_begin?: string | null;
  date_end?: string | null;
  check_point: number;
  img_path?: string | null;
}

export interface PaginatedResponse<T> {
  data: T[];
  pagination: {
    current_page: number;
    limit: number;
    total_count: number;
    total_pages: number;
  };
}